### PR TITLE
[Refactor]: Support batch infer in some model

### PR DESCRIPTION
**Pull request by hhaAndroid**
**Environment**:
bj17, TITANXP@12G

**Modified model**：
RetinaNet/YOLOV3/FCOS/ATSS/PAA/RPN.

**Precision**:
All model's precision is consistent. The model list is as follows

**FPS**:
`tools/analysis_tools/benchmark.py`

batch=1, master, self:  

```sh
  "configs/atss/atss_r50_fpn_1x_coco.py": 16.4, 16.3
  "configs/carafe/faster_rcnn_r50_fpn_carafe_1x_coco.py": 13.9,14.0
  "configs/cascade_rcnn/cascade_rcnn_r50_fpn_1x_coco.py": 11.0,11.0
  "configs/cascade_rcnn/cascade_mask_rcnn_r50_fpn_1x_coco.py": 8.0,7.8
  "configs/cascade_rpn/crpn_r50_caffe_fpn_1x_coco.py": 11.8,11.7
  "configs/centripetalnet/centripetalnet_hourglass104_mstest_16x6_210e_coco.py": 3.4,3.4,
  "configs/cornernet/cornernet_hourglass104_mstest_8x6_210e_coco.py": 4.2,4.2,
  "configs/dcn/faster_rcnn_r50_fpn_dconv_c3-c5_1x_coco.py": 13.6,13.4,
  "configs/detectors/detectors_htc_r50_1x_coco.py": 3.0,3.0,
  "configs/detr/detr_r50_8x2_150e_coco.py": 18.4, 18.2,
  "configs/double_heads/dh_faster_rcnn_r50_fpn_1x_coco.py": 6.7,6.7,
  "configs/dynamic_rcnn/dynamic_rcnn_r50_fpn_1x.py": 15.0,15.1,
  "configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco.py": 14.9,15.2,
  "configs/fcos/fcos_center-normbbox-centeronreg-giou_r50_caffe_fpn_gn-head_1x_coco.py": 18.2,18.3,
  "configs/foveabox/fovea_align_r50_fpn_gn-head_4x4_2x_coco.py": 14.9,14.9,
  "configs/free_anchor/retinanet_free_anchor_r50_fpn_1x_coco.py": 13.6,13.6,
  "configs/fsaf/fsaf_r50_fpn_1x_coco.py": 17.4,17.5,
  "configs/gfl/gfl_r50_fpn_1x_coco.py": 16.7,16.8,
  "configs/grid_rcnn/grid_rcnn_r50_fpn_gn-head_2x_coco.py": 11.8,11.8,
  "configs/guided_anchoring/ga_retinanet_r50_caffe_fpn_1x_coco.py": 12.8,12.6,
  "configs/guided_anchoring/ga_faster_r50_caffe_fpn_1x_coco.py": 11.4,11.3,
  "configs/htc/htc_r50_fpn_1x_coco.py": 4.1, 4.0,
  "configs/mask_rcnn/mask_rcnn_r50_fpn_1x_coco.py": 13.0,13.2,
  "configs/nas_fpn/retinanet_r50_nasfpn_crop640_50e_coco.py": 16.2,16.1,
  "configs/paa/paa_r50_fpn_1x_coco.py": 4.4,4.5,
  "configs/pisa/pisa_faster_rcnn_r50_fpn_1x_coco.py": 13.4,13.1,
  "configs/point_rend/point_rend_r50_caffe_fpn_mstrain_1x_coco.py": 8.6,8.5,
  "configs/reppoints/reppoints_moment_r50_fpn_1x_coco.py": 13.8,13.8,
  "configs/retinanet/retinanet_r50_fpn_1x_coco.py": 15.2, 15.0,
  "configs/rpn/rpn_r50_fpn_1x_coco.py": 18.7, 18.7,
  "configs/sabl/sabl_faster_rcnn_r50_fpn_1x_coco.py": 12.8,12.8,
  "configs/sabl/sabl_retinanet_r50_fpn_1x_coco.py": 14.6,14.6,
  "configs/scnet/scnet_r50_fpn_1x_coco.py": 4.5,4.5,
  "configs/sparse_rcnn/sparse_rcnn_r50_fpn_1x_coco.py": 14.1,14.3,
  "configs/ssd/ssd300_coco.py": 51.7,51.2,
  "configs/tridentnet/tridentnet_r50_caffe_1x_coco.py": 2.3, 2.2,
  "configs/vfnet/vfnet_r50_fpn_1x_coco.py": 14.0,13.9,
  "configs/yolact/yolact_r50_1x8_coco.py": 25.2,27.3,
  "configs/yolo/yolov3_d53_320_273e_coco.py": 42.7，41.3
 ```

batch=8, master, self:  
```sh
 "configs/atss/atss_r50_fpn_1x_coco.py": 1.631,1.574
  "configs/carafe/faster_rcnn_r50_fpn_carafe_1x_coco.py": 1.38,1.352
  "configs/cascade_rcnn/cascade_rcnn_r50_fpn_1x_coco.py": 1.164,1.151
  "configs/cascade_rcnn/cascade_mask_rcnn_r50_fpn_1x_coco.py": 0.878,0.905
  "configs/dcn/faster_rcnn_r50_fpn_dconv_c3-c5_1x_coco.py": 1.297,1.284
  "configs/detectors/detectors_htc_r50_1x_coco.py": 0.345,0.343
  "configs/dynamic_rcnn/dynamic_rcnn_r50_fpn_1x.py": 1.441,1.434
  "configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco.py": 1.467,1.429
  "configs/fcos/fcos_center-normbbox-centeronreg-giou_r50_caffe_fpn_gn-head_1x_coco.py": 1.739,1.699
  "configs/foveabox/fovea_align_r50_fpn_gn-head_4x4_2x_coco.py": 1.341,1.333
  "configs/free_anchor/retinanet_free_anchor_r50_fpn_1x_coco.py": 1.386,1.367
  "configs/fsaf/fsaf_r50_fpn_1x_coco.py": 1.719,1.71
  "configs/gfl/gfl_r50_fpn_1x_coco.py": 1.661,1.631
  "configs/grid_rcnn/grid_rcnn_r50_fpn_gn-head_2x_coco.py": 1.32,1.288
  "configs/htc/htc_r50_fpn_1x_coco.py": 0.447,0.47
  "configs/mask_rcnn/mask_rcnn_r50_fpn_1x_coco.py": 1.316,1.318
  "configs/nas_fpn/retinanet_r50_nasfpn_crop640_50e_coco.py": 3.325,3.316
  "configs/paa/paa_r50_fpn_1x_coco.py": 0.526,0.523
  "configs/pisa/pisa_faster_rcnn_r50_fpn_1x_coco.py": 1.313,1.298
  "configs/point_rend/point_rend_r50_caffe_fpn_mstrain_1x_coco.py": 0.954,0.947
  "configs/reppoints/reppoints_moment_r50_fpn_1x_coco.py": 1.473,1.418
  "configs/retinanet/retinanet_r50_fpn_1x_coco.py": 1.493,1.45
  "configs/rpn/rpn_r50_fpn_1x_coco.py": 1.713, 1.669
  "configs/sabl/sabl_faster_rcnn_r50_fpn_1x_coco.py": 1.284,1.255
  "configs/sabl/sabl_retinanet_r50_fpn_1x_coco.py": 1.56,1.53
  "configs/scnet/scnet_r50_fpn_1x_coco.py": 0.555,0.547
  "configs/sparse_rcnn/sparse_rcnn_r50_fpn_1x_coco.py": 1.732,1.678
  "configs/ssd/ssd300_coco.py": 8.339,8.235
  "configs/vfnet/vfnet_r50_fpn_1x_coco.py": 1.467,1.424
  "configs/yolact/yolact_r50_1x8_coco.py": 4.219,4.105
  "configs/yolo/yolov3_d53_320_273e_coco.py": 11.823,11.509
```

### CODE REVIEW

**Comment by hhaAndroid** [comment]
local environment:

- 15.6 GiB
- Intel® Core™ i7-8700 CPU @ 3.20GHz × 12
- GeForce GTX 1060 6GB/PCIe/SSE2

`python tools/analysis_tools/benchmark.py configs/retinanet/retinanet_r50_fpn_1x_coco.py retinanet_r50_fpn_1x_coco_20200130-c2398f9e.pth 
`

before 
      size    |      batch   |  before speed (img/s) | after speed (img/s)
 - (1333,800)  |  1 |  6.8 | 6.8
 - (1333,800)  |  2 |  3.1x2=6.2 | 3.1x2=6.2
 - (1333,800)  |  4 |  1.4x4=5.6 | 1.4x4=5.6
  - 
 - (600,400)  |  1 |  19.3 | 18.5
 - (600,400)  |  2 |  9.9x2=19.8 | 9.7x2=19.4
 - (600,400)  |  4 |  4.7x4=18.8 | 4.8x4=19.2
 - 
 - (400,300)  |  1 |  27.2 | 26.4
 - (400,300)  |  2 |  15.4x2=30.8 | 15.8x2=31.6
 - (400,300) |  4 |  7.8x4=31.2 | 8.6x4=34.4

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 1]
```diff
@@ -171,22 +171,29 @@ def delta2bbox(rois,
...
+    dh = denorm_deltas[..., 3::4]
     max_ratio = np.abs(np.log(wh_ratio_clip))
     dw = dw.clamp(min=-max_ratio, max=max_ratio)
     dh = dh.clamp(min=-max_ratio, max=max_ratio)
+    x1, y1 = rois[..., 0], rois[..., 1]
+    x2, y2 = rois[..., 2], rois[..., 3]
     # Compute center of each roi
-    px = ((rois[:, 0] + rois[:, 2]) * 0.5).unsqueeze(1).expand_as(dx)
-    py = ((rois[:, 1] + rois[:, 3]) * 0.5).unsqueeze(1).expand_as(dy)
+    px = ((x1 + x2) * 0.5).unsqueeze(2).expand_as(dx)
```
Comment (ZwwWayne) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:192]: should unsqueeze(-1)

File: mmdet/models/dense_heads/anchor_head.py [hunk 2]
```diff
@@ -620,70 +618,89 @@ def _get_bboxes_single(self,
         """
         cfg = self.test_cfg if cfg is None else cfg
         assert len(cls_score_list) == len(bbox_pred_list) == len(mlvl_anchors)
+        batch_size = cls_score_list[0].shape[0]
         # convert to tensor to keep tracing
         nms_pre_tensor = torch.tensor(
```
Comment (jshilong) [inline mmdet/models/dense_heads/anchor_head.py:623]: Seems `cls_score_list[0].new_tensor` would be ok. Seems that `dtype=torch.long` is not necessary ?

`

File: mmdet/models/dense_heads/anchor_head.py [hunk 3]
```diff
@@ -620,70 +618,89 @@ def _get_bboxes_single(self,
...
-                                          0).reshape(-1, self.cls_out_channels)
+            cls_score = cls_score.permute(0, 2, 3,
+                                          1).reshape(batch_size, -1,
+                                                     self.cls_out_channels)
             if self.use_sigmoid_cls:
                 scores = cls_score.sigmoid()
             else:
                 scores = cls_score.softmax(-1)
-            bbox_pred = bbox_pred.permute(1, 2, 0).reshape(-1, 4)
+            bbox_pred = bbox_pred.permute(0, 2, 3,
```
Comment (jshilong) [inline mmdet/models/dense_heads/anchor_head.py:640]: This is contradictory to  `deltas (Tensor): Encoded offsets with respect to each roi has shape (N, 4 * num_classes) or (N, 4).`

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 4]
```diff
@@ -198,10 +208,18 @@ def delta2bbox(rois,
...
-    if clip_border and max_shape is not None:
-        x1 = x1.clamp(min=0, max=max_shape[1])
-        y1 = y1.clamp(min=0, max=max_shape[0])
-        x2 = x2.clamp(min=0, max=max_shape[1])
-        y2 = y2.clamp(min=0, max=max_shape[0])
+
     bboxes = torch.stack([x1, y1, x2, y2], dim=-1).view(deltas.size())
+
+    if clip_border and max_shape is not None:
+        if isinstance(max_shape, list):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:215]: Add comments indicating why do not use clamp.

File: mmdet/models/dense_heads/anchor_head.py [hunk 5]
```diff
@@ -620,70 +618,96 @@ def _get_bboxes_single(self,
...
-                num_anchor = _shape_as_tensor(scores)[-2]
+                num_anchor = _shape_as_tensor(scores)[-2].to(
+                    nms_pre_tensor.device)
                 nms_pre = torch.where(nms_pre_tensor < num_anchor,
                                       nms_pre_tensor, num_anchor)
+
                 # Get maximum scores for foreground classes.
                 if self.use_sigmoid_cls:
-                    max_scores, _ = scores.max(dim=1)
+                    max_scores, _ = scores.max(dim=2)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/anchor_head.py:655]: May use -1?

**Review by ZwwWayne — APPROVED** [review]

**Comment by ZwwWayne** [comment]
This PR seems to be very big. Can we split it into two PR? The first one only modifies the box encoding & decoding process. The second one refactor heads to support ONNX exportable batch inference.

**Comment by hhaAndroid** [comment]
> This PR seems to be very big. Can we split it into two PR? The first one only modifies the box encoding & decoding process. The second one refactor heads to support ONNX exportable batch inference.

good idea

File: mmdet/models/dense_heads/yolo_head.py [hunk 6]
```diff
@@ -230,106 +228,143 @@ def _get_bboxes_single(self,
...
-                # add as_tuple=False for compatibility in Pytorch 1.6
-                # flatten would create a Reshape op with constant values,
-                # and raise RuntimeError when doing inference in ONNX Runtime
-                # with a different input image (#4221).
-                conf_inds = conf_pred.ge(conf_thr).nonzero(
-                    as_tuple=False).squeeze(1)
-                bbox_pred = bbox_pred[conf_inds, :]
-                cls_pred = cls_pred[conf_inds, :]
-                conf_pred = conf_pred[conf_inds]
+            # conf_thr = cfg.get('conf_thr', -1)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolo_head.py:283]: Clean commented code.

**Comment by RunningLeon** [comment]
Tests on pytorch2onnx
There are four kinds of deployment policies:

- Option 1: batch_size=1 and model input shape is fixed.
- Option 2: batch_size=1 and model input shape is dynamic.
- Option 3: batch_size>1 and model input shape is fixed.
- Option 4: batch_size>1 and model input shape is dynamic.

The table below lists the models that are guaranteed to be exportable to ONNX and runnable in ONNX Runtime.

|    Model    |                          Config                          | Option 1 | Option 2 | Option 3 | Option 4 |
| :---------: | :------------------------------------------------------: | :------: | :------: | :------: | :------: |
|     SSD     |               `configs/ssd/ssd300_coco.py`               |    Y     |    Y     | Feasible | Feasible |
|   YOLOv3    |    `configs/yolo/yolov3_d53_mstrain-608_273e_coco.py`    |    Y     |    Y     | Feasible | Feasible |
|    FSAF     |          `configs/fsaf/fsaf_r50_fpn_1x_coco.py`          |    Y     |    Y     | Feasible | Feasible |
|  RetinaNet  |     `configs/retinanet/retinanet_r50_fpn_1x_coco.py`     |    Y     |    Y     | Feasible | Feasible |
|    FCOS     | `configs/fcos/fcos_r50_caffe_fpn_gn-head_4x4_1x_coco.py` |    Y     |    Y     | Feasible | Feasible |
| Faster-RCNN |   `configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco.py`   |    Y     |    Y     |    TBD     |    TBD    |

File: mmdet/models/dense_heads/paa_head.py [hunk 7]
```diff
@@ -516,76 +516,92 @@ def _get_targets_single(self,
...
-                           cls_scores,
-                           bbox_preds,
-                           iou_preds,
-                           mlvl_anchors,
-                           img_shape,
-                           scale_factor,
-                           cfg,
-                           rescale=False,
-                           with_nms=True):
+    def _get_bboxes(self,
```
Comment (jshilong) [inline mmdet/models/dense_heads/paa_head.py:519]: 1 comments the length of the list
2 (num_points * num_classes, H, W) in docstr maybe wrong?

File: mmdet/models/dense_heads/rpn_head.py [hunk 8]
```diff
@@ -79,14 +79,14 @@ def loss(self,
...
 
-    def _get_bboxes_single(self,
-                           cls_scores,
-                           bbox_preds,
-                           mlvl_anchors,
-                           img_shape,
-                           scale_factor,
-                           cfg,
-                           rescale=False):
+    def _get_bboxes(self,
```
Comment (jshilong) [inline mmdet/models/dense_heads/rpn_head.py:82]: 1 comments the length of the list
2 (num_points * num_classes, H, W) in docstr maybe wrong?

File: mmdet/models/dense_heads/yolo_head.py [hunk 9]
```diff
@@ -197,30 +197,28 @@ def get_bboxes(self,
...
+                                       rescale, with_nms)
         return result_list
 
-    def _get_bboxes_single(self,
-                           pred_maps_list,
-                           scale_factor,
-                           cfg,
-                           rescale=False,
-                           with_nms=True):
+    def _get_bboxes(self,
```
Comment (jshilong) [inline mmdet/models/dense_heads/yolo_head.py:210]: 1 comments the length of the list
2 (num_points * num_classes, H, W) in docstr maybe wrong?

File: tests/test_utils/test_misc.py [hunk 10]
```diff
@@ -45,3 +46,42 @@ def test_mask2ndarray():
...
+    assert out.allclose(batch_out)
+
+    # test max_shape is not equal to batch
+    with pytest.raises(AssertionError):
+        distance2bbox(
+            batch_point,
+            batch_distance,
+            max_shape=[(120, 100), (120, 100), (32, 32)])
+
+    rois = torch.zeros((0, 4))
```
Comment (jshilong) [inline tests/test_utils/test_misc.py:84]: test B × 0 × 4

**Review by jshilong — APPROVED** [review]

**Pull request merged**